[BUG] with_hidden_metadata_fallback test failures on Databricks 11.3

test_csv_scan_with_hidden_metadata_fallback, test_orc_scan_with_hidden_metadata_fallback, and test_parquet_scan_with_hidden_metadata_fallback are all failing on Databricks 11.3 with ArrayIndexOutOfBoundsException.  It looks like the query plan is failing to fallback to the CPU and is crashing attempting to access a metadata column that is not present.

For example:
```
__________ test_parquet_scan_with_hidden_metadata_fallback[file_size] __________

spark_tmp_path = '/tmp/pyspark_tests//1129-193707-fzot5ykw-10-59-240-236-master-3884-1602655101/'
metadata_column = 'file_size'

    @pytest.mark.skipif(is_before_spark_330(), reason='Hidden file metadata columns are a new feature of Spark 330')
    @allow_non_gpu(any = True)
    @pytest.mark.parametrize('metadata_column', ["file_path", "file_name", "file_size", "file_modification_time"])
    def test_parquet_scan_with_hidden_metadata_fallback(spark_tmp_path, metadata_column):
        data_path = spark_tmp_path + "/hidden_metadata.parquet"
        with_cpu_session(lambda spark : spark.range(10) \
                         .selectExpr("id", "id % 3 as p") \
                         .write \
                         .partitionBy("p") \
                         .mode("overwrite") \
                         .parquet(data_path))
    
        def do_parquet_scan(spark):
            df = spark.read.parquet(data_path).selectExpr("id", "_metadata.{}".format(metadata_column))
            return df
    
>       assert_cpu_and_gpu_are_equal_collect_with_capture(
            do_parquet_scan,
            exist_classes= "FileSourceScanExec",
            non_exist_classes= "GpuBatchScanExec")

../../src/main/python/parquet_test.py:927: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../src/main/python/asserts.py:369: in assert_cpu_and_gpu_are_equal_collect_with_capture
    from_gpu, gpu_df = with_gpu_session(bring_back, conf=conf)
../../src/main/python/spark_session.py:130: in with_gpu_session
    return with_spark_session(func, conf=copy)
../../src/main/python/spark_session.py:97: in with_spark_session
    ret = func(_spark)
../../src/main/python/asserts.py:209: in bring_back
    return (df.collect(), df)
/databricks/spark/python/pyspark/instrumentation_utils.py:48: in wrapper
    res = func(*args, **kwargs)
/databricks/spark/python/pyspark/sql/dataframe.py:837: in collect
    sock_info = self._jdf.collectToPython()
/databricks/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py:1321: in __call__
    return_value = get_return_value(
/databricks/spark/python/pyspark/sql/utils.py:196: in deco
    return f(*a, **kw)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

answer = 'xro2703'
gateway_client = <py4j.clientserver.JavaClient object at 0x7f8eb9504b80>
target_id = 'o2700', name = 'collectToPython'

    def get_return_value(answer, gateway_client, target_id=None, name=None):
        """Converts an answer received from the Java gateway into a Python object.
    
        For example, string representation of integers are converted to Python
        integer, string representation of objects are converted to JavaObject
        instances, etc.
    
        :param answer: the string returned by the Java gateway
        :param gateway_client: the gateway client used to communicate with the Java
            Gateway. Only necessary if the answer is a reference (e.g., object,
            list, map)
        :param target_id: the name of the object from which the answer comes from
            (e.g., *object1* in `object1.hello()`). Optional.
        :param name: the name of the member from which the answer comes from
            (e.g., *hello* in `object1.hello()`). Optional.
        """
        if is_error(answer)[0]:
            if len(answer) > 1:
                type = answer[1]
                value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
                if answer[1] == REFERENCE_TYPE:
>                   raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
                        format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o2700.collectToPython.
E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 2 in stage 46.0 failed 1 times, most recent failure: Lost task 2.0 in stage 46.0 (TID 144) (ip-10-59-240-236.us-west-2.compute.internal executor driver): java.lang.ArrayIndexOutOfBoundsException: 2
E                   	at org.apache.spark.sql.vectorized.ColumnarBatch.column(ColumnarBatch.java:98)
E                   	at com.nvidia.spark.rapids.GpuBoundReference.columnarEval(GpuBoundAttribute.scala:157)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
E                   	at com.nvidia.spark.rapids.GpuExpressionsUtils$.columnarEvalToColumn(GpuExpressions.scala:94)
E                   	at org.apache.spark.sql.rapids.GpuCreateNamedStruct.$anonfun$columnarEval$9(complexTypeCreator.scala:231)
E                   	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
E                   	at org.apache.spark.sql.rapids.GpuCreateNamedStruct.$anonfun$columnarEval$8(complexTypeCreator.scala:230)
E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:55)
E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:53)
E                   	at org.apache.spark.sql.rapids.GpuCreateNamedStruct.withResource(complexTypeCreator.scala:164)
E                   	at org.apache.spark.sql.rapids.GpuCreateNamedStruct.columnarEval(complexTypeCreator.scala:228)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
E                   	at com.nvidia.spark.rapids.GpuAlias.columnarEval(namedExpressions.scala:109)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
E                   	at com.nvidia.spark.rapids.GpuExpressionsUtils$.columnarEvalToColumn(GpuExpressions.scala:94)
E                   	at com.nvidia.spark.rapids.GpuProjectExec$.projectSingle(basicPhysicalOperators.scala:108)
E                   	at com.nvidia.spark.rapids.GpuProjectExec$.$anonfun$project$1(basicPhysicalOperators.scala:115)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$MapsSafely.$anonfun$safeMap$1(implicits.scala:216)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$MapsSafely.$anonfun$safeMap$1$adapted(implicits.scala:213)
E                   	at scala.collection.immutable.List.foreach(List.scala:431)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$MapsSafely.safeMap(implicits.scala:213)
E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$AutoCloseableProducingSeq.safeMap(implicits.scala:248)
E                   	at com.nvidia.spark.rapids.GpuProjectExec$.project(basicPhysicalOperators.scala:115)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$tieredProject$1(basicPhysicalOperators.scala:335)
E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.withResource(basicPhysicalOperators.scala:286)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.recurse$1(basicPhysicalOperators.scala:334)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.tieredProject(basicPhysicalOperators.scala:354)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$tieredProjectAndClose$2(basicPhysicalOperators.scala:360)
E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.withResource(basicPhysicalOperators.scala:286)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$tieredProjectAndClose$1(basicPhysicalOperators.scala:359)
E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.withResource(basicPhysicalOperators.scala:286)
E                   	at com.nvidia.spark.rapids.GpuTieredProject.tieredProjectAndClose(basicPhysicalOperators.scala:358)
E                   	at com.nvidia.spark.rapids.GpuProjectExec.$anonfun$doExecuteColumnar$1(basicPhysicalOperators.scala:161)
E                   	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
E                   	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$2(GpuColumnarToRowExec.scala:241)
E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.withResource(GpuColumnarToRowExec.scala:187)
E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:238)
E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:215)
E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:255)
E                   	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
E                   	at org.apache.spark.sql.execution.collect.UnsafeRowBatchUtils$.encodeUnsafeRows(UnsafeRowBatchUtils.scala:80)
E                   	at org.apache.spark.sql.execution.collect.Collector.$anonfun$processFunc$1(Collector.scala:186)
E                   	at org.apache.spark.scheduler.ResultTask.$anonfun$runTask$3(ResultTask.scala:75)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.scheduler.ResultTask.$anonfun$runTask$1(ResultTask.scala:75)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:55)
E                   	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:169)
E                   	at org.apache.spark.scheduler.Task.$anonfun$run$4(Task.scala:137)
E                   	at com.databricks.unity.EmptyHandle$.runWithAndClose(UCSHandle.scala:104)
E                   	at org.apache.spark.scheduler.Task.$anonfun$run$1(Task.scala:137)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:96)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$13(Executor.scala:902)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1696)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:905)
E                   	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E                   	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:760)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E                   	at java.lang.Thread.run(Thread.java:750)
```

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/Spark321PlusDBShims.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/Spark321PlusDBShims.scala
@@ -106,6 +106,10 @@ trait Spark321PlusDBShims extends SparkShims
     }
   }
 
+  def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
+    GpuFileSourceScanExec.tagSupport(meta)
+  }
+
   private val shimExecs: Map[Class[_ <: SparkPlan], ExecRule[_ <: SparkPlan]] = {
     Seq(
       GpuOverrides.exec[SubqueryBroadcastExec](
@@ -194,6 +198,8 @@ trait Spark321PlusDBShims extends SparkShims
           override def convertToCpu(): SparkPlan = {
             wrapped.copy(partitionFilters = partitionFilters)
           }
+
+          override def tagPlanForGpu(): Unit = tagFileSourceScanExec(this)
 
           override def convertToGpu(): GpuExec = {
             val sparkSession = wrapped.relation.sparkSession

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/Spark321PlusDBShims.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/Spark321PlusDBShims.scala
@@ -106,10 +106,6 @@ trait Spark321PlusDBShims extends SparkShims
     }
   }
 
-  def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
-    GpuFileSourceScanExec.tagSupport(meta)
-  }
-
   private val shimExecs: Map[Class[_ <: SparkPlan], ExecRule[_ <: SparkPlan]] = {
     Seq(
       GpuOverrides.exec[SubqueryBroadcastExec](

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/Spark321PlusDBShims.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/Spark321PlusDBShims.scala
@@ -192,14 +192,12 @@ trait Spark321PlusDBShims extends SparkShims
               this.entirePlanWillNotWork("Plans that read Delta Index JSON files can not run " +
                   "any part of the plan on the GPU!")
             }
-            GpuFileSourceScanExec.tagSupport(this)
+            tagFileSourceScanExec(this)
           }
 
           override def convertToCpu(): SparkPlan = {
             wrapped.copy(partitionFilters = partitionFilters)
           }
-
-          override def tagPlanForGpu(): Unit = tagFileSourceScanExec(this)
 
           override def convertToGpu(): GpuExec = {
             val sparkSession = wrapped.relation.sparkSession

--- a/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
+++ b/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
@@ -46,6 +46,7 @@ object SparkShimImpl extends Spark321PlusDBShims {
   }
 
   override def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
+    println("hit the shim");
     if (meta.wrapped.expressions.exists {
       case FileSourceMetadataAttribute(_) => true
       case _ => false

--- a/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
+++ b/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
@@ -46,7 +46,7 @@ object SparkShimImpl extends Spark321PlusDBShims {
   }
 
   override def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
-    println("hit the shim");
+    throw new ArithmeticException("HIT THE SHIM OF DB113 !!!!!")
     if (meta.wrapped.expressions.exists {
       case FileSourceMetadataAttribute(_) => true
       case _ => false

--- a/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
+++ b/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
@@ -45,6 +45,16 @@ object SparkShimImpl extends Spark321PlusDBShims {
       pushDownInFilterThreshold, caseSensitive, datetimeRebaseMode)
   }
 
+  override def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
+    if (meta.wrapped.expressions.exists {
+      case FileSourceMetadataAttribute(_) => true
+      case _ => false
+    }) {
+      meta.willNotWorkOnGpu("hidden metadata columns are not supported on GPU")
+    }
+    super.tagFileSourceScanExec(meta)
+  }
+
   override def getExprs: Map[Class[_ <: Expression], ExprRule[_ <: Expression]] =
     super.getExprs ++ DayTimeIntervalShims.exprs ++ RoundingShims.exprs
 

--- a/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
+++ b/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
@@ -21,7 +21,7 @@ import org.apache.parquet.schema.MessageType
 
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.trees.TreePattern._
-import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.{FileSourceScanExec, SparkPlan}
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFilters
 

--- a/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
+++ b/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
@@ -45,7 +45,7 @@ object SparkShimImpl extends Spark321PlusDBShims {
       pushDownInFilterThreshold, caseSensitive, datetimeRebaseMode)
   }
 
-  def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
+  override def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
     if (meta.wrapped.expressions.exists {
       case FileSourceMetadataAttribute(_) => true
       case _ => false

--- a/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
+++ b/sql-plugin/src/main/330db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala
@@ -45,8 +45,7 @@ object SparkShimImpl extends Spark321PlusDBShims {
       pushDownInFilterThreshold, caseSensitive, datetimeRebaseMode)
   }
 
-  override def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
-    throw new ArithmeticException("HIT THE SHIM OF DB113 !!!!!")
+  def tagFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
     if (meta.wrapped.expressions.exists {
       case FileSourceMetadataAttribute(_) => true
       case _ => false

